contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+is to allow administrators to request fail open on specific calls, or to require that certain special resource
+paths (initializers dynamic config path) are always fail open. Alternatively, it may be desirable for
+administrative users to bypass admission completely.
+
+Some options:
+
+1. Some namespaces are opted out of external admission (kube-system)
+2. Certain service accounts can bypass external admission checks
+
+
+##### Upgrade of a cluster with external admission
+
+The current order of cluster upgrade is `apiserver` -> `controller` -> `nodes`.  External admission controllers
+would typically need to be upgraded first in order to ensure new semantic changes in objects are not ignored.
+This would include fields like PodSecurityContext - adding that *prior* to admission is necessary because it
+allows escalation that was previously impossible.
+
+
+#### Dynamic configuration
+
+Trusted clients should be able to modify the initialization and external admission hooks on the fly and expect
+that configuration is updated quickly. Extension API servers should also be able to leverage central
+configuration, but may opt for alternate mechanisms.
+
+The initializer admission controller and the generic webhook admission controller should dynamically load config
+from a `ConfigMap` or a net new API object holding the following configuration schema:
+
+```
+type AdmissionControlConfiguration struct {
+    TypeMeta // although this object could simply be serialized like ComponentConfig
+
+    // ResourceInitializers is a list of resources and their default initializers
+    ResourceInitializers []ResourceDefaultInitializer
+
+    ExternalAdmissionHooks []ExternalAdmissionHook
+}
+
+type ResourceDefaultInitializer struct {
+    // Resource identifies the type of resource to be initialized that should be initialized
+    Resource GroupResource
+    // Initializers are the default names that will be registered to this resource
+    Initializers []string
+}
+
+type ExternalAdmissionHook struct {
+    // Operations is the list of operations this hook will be invoked on - Create, Update, or *
+    // for all operations. Defaults to '*'.
+    Operations []string
+    // Resources are the resources this hook should be invoked on. '*' is all resources.
+    Resources []string
+    // Subresources are the list of subresources this hook should be invoked on. '*' is all resources.
+    Subresources []string
+
+    // TODO define client configuration
+
+    // FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+    // allowed values are Ignore, Retry, Fail. Default value is Fail
+    FailurePolicy FailurePolicyType
+}
+```
+
+All changes to this config must be done in an safe matter - when adding a new hook or initializer,
+first verify the new agent is online before allowing it to come into rotation. Removing a hook or
+initializer must occur before disabling the remote endpoint, and all queued items must complete.
```

Comment:
@smarterclayton "other apiservers can be arbitrarily delayed" are you assuming this is going to be cached by apiservers?

I think basically the case where a controller goes unavailable HAS to be handled at run time. We can check it at add-time but that doesn't help if the controller goes offline later, or folds under load, etc. The person adding the initalizer may want to decide if it's appropriate to fail-open or fail-closed. If we're providing the choice, then perhaps instead of checking at add time, we create these things in fail-open mode and--if they prove stable after some number of invocations--upgrade them to fail-closed mode.